### PR TITLE
BtreeMap: refactoring around edges

### PR DESCRIPTION
**Pull request by ssomers**
Parts chipped off a more daring effort, that the btree benchmarks judge to be performance-neutral.

r? @Mark-Simulacrum

### CODE REVIEW

File: library/alloc/src/collections/btree/node.rs [hunk 1]
```diff
@@ -483,18 +494,49 @@ impl<'a, K, V, Type> NodeRef<marker::Mut<'a>, K, V, Type> {
         unsafe { self.reborrow_mut().into_val_mut_at(idx) }
     }
 
-    fn keys_mut(&mut self) -> &mut [K] {
+    fn keys_mut(&mut self) -> &mut [K]
+    where
+        K: 'a,
+        V: 'a,
```
Comment (ssomers) [inline library/alloc/src/collections/btree/node.rs:500]: The need for this bound is created by inlining into_key_slice_mut, but don't ask me why.

**Comment by ssomers** [comment]
Rebased & some more minor and comment tweaks added

File: library/alloc/src/collections/btree/node.rs [hunk 2]
```diff
@@ -47,8 +47,7 @@ const KV_IDX_CENTER: usize = B - 1;
 const EDGE_IDX_LEFT_OF_CENTER: usize = B - 1;
 const EDGE_IDX_RIGHT_OF_CENTER: usize = B;
 
-/// The underlying representation of leaf nodes.
-#[repr(C)]
```
Comment (Mark-Simulacrum) [inline library/alloc/src/collections/btree/node.rs:51]: Why was this repr(C) removed? I'm pretty confident it is indeed still necessary for soundness of treating LeafNode as a prefix of InternalNode.

Comment (ssomers) [reply library/alloc/src/collections/btree/node.rs:51]: Well, I'm pretty confident it is not. The introduction of the empty root node (#50352) made it necessary, but you simplified that away, so we don't need to constrain leaf layout. For a particular K and V, whatever layout the compiler chooses , the field in `InternalNode` has the same layout and the repr(C) on that guarantees that the internal node's alignment is at least as coarse as the leaf node's. The unit test `test_iter_mut_mutation` covers alignment of internal nodes. And from voluntary and involuntary experiments, I can tell the compiler is eager to reorder fields if you let your guard down.

Comment (Mark-Simulacrum) [reply library/alloc/src/collections/btree/node.rs:51]: Yes, thinking more, that's correct -- if we want to take a prefix of LeafNode and treat that specially (e.g., assume parent is the first thing) then that would be different, but we don't do that.

**Comment by Mark-Simulacrum** [comment]
@bors r+

Thanks!

**Pull request merged**